packages/altair-api/src/auth/user/user.service.spec.ts

```diff
@@ -2,19 +2,230 @@ import { Test, TestingModule } from '@nestjs/testing';
 import { PrismaService } from 'nestjs-prisma';
 import { StripeService } from 'src/stripe/stripe.service';
 import { UserService } from './user.service';
+import { SignupInput } from '../models/signup.input';
+import {
+  mockPlanConfig,
+  mockPrismaConflictError,
+  mockUser,
+  mockUserPlan,
+} from '../mocks/prisma-service.mock';
+import {
+  mockStripeCustomer,
+  mockSubscriptionItem,
+} from '../mocks/stripe-service.mock';
+import Stripe from 'stripe';
 
 describe('UserService', () => {
   let service: UserService;
+  let stripeService: StripeService;
+  let prismaService: PrismaService;
 
   beforeEach(async () => {
     const module: TestingModule = await Test.createTestingModule({
       providers: [UserService, PrismaService, StripeService],
     }).compile();
 
     service = module.get<UserService>(UserService);
+    stripeService = module.get<StripeService>(StripeService);
+    prismaService = module.get<PrismaService>(PrismaService);
   });
 
   it('should be defined', () => {
     expect(service).toBeDefined();
   });
+
+  describe('createUser', () => {
+    let payloadMock: SignupInput;
+
+    beforeAll(() => {
+      payloadMock = { email: 'john.doe@email.com' };
+    });
+
+    beforeEach(() => {
+      jest
+        .spyOn(stripeService, 'connectOrCreateCustomer')
+        .mockResolvedValueOnce(mockStripeCustomer());
+    });
+
+    it('should return the created user object', async () => {
+      // GIVEN
+      jest
+        .spyOn(prismaService.user, 'create')
+        .mockResolvedValueOnce(mockUser());
+
+      // WHEN
+      const user = await service.createUser(payloadMock);
+
+      // THEN
+      (expect(user) as any).toBeUser();
+    });
+
+    it('should throw and exception if the email is already taken', () => {
+      // GIVEN
+      jest
+        .spyOn(prismaService.user, 'create')
+        .mockRejectedValueOnce(mockPrismaConflictError());
+
+      // THEN
+      expect(service.createUser(payloadMock)).rejects.toThrow(
+        'Email john.doe@email.com already used.'
+      );
+    });
+
+    it('should rethrow an unknown error', () => {
+      // GIVEN
+      jest
+        .spyOn(prismaService.user, 'create')
+        .mockRejectedValueOnce(Error('Unexpected error'));
+
+      // THEN
+      expect(service.createUser(payloadMock)).rejects.toThrow(
+        'Unexpected error'
+      );
+    });
+  });
+
+  describe('mustGetUser', () => {
+    it('should return a user object', async () => {
+      // GIVEN
+      const userMock = mockUser();
+      jest.spyOn(service, 'getUser').mockResolvedValueOnce(userMock);
+
+      // WHEN
+      const user = await service.mustGetUser(userMock.id);
+
+      // THEN
+      (expect(user) as any).toBeUser();
```

Comment:
I like that you have implemented custom matchers. Can we add typings for the custom matchers also, instead of using `any`? It would also help with discovery so other authors know it exists?

https://redd.one/blog/practical-guide-to-custom-jest-matchers

Reply:
@imolorhe can you give me some guidance here, please?

I defined the type extension (jest.d.ts) under the `altair-api` folder. But I'm having issues with which `tsconfig.json` to include.

This is what I have tried already:
1. The package's (`packages/altair-api/tsconfig.json`)
2. The root (`tsconfig.json`)
3. Create a spec tsconfig under the package (`packages/altair-api/tsconfig.spec.json`) and point the jest config to it.

The test suite keeps failing on the missing type:

`src/auth/user/user.service.spec.ts:60:20 - error TS2339: Property 'toBeUser' does not exist on type 'JestMatchers<User>'.`

Reply:
I think what you'll want is to create it under `@types/jest/index.d.ts` (similar to the one for [express](https://github.com/altair-graphql/altair/blob/master/packages/altair-api/%40types/express/index.d.ts))

Reply:
I tried placing it under the `@types` folder both as `@types/jest/index.d.ts` and `@types/jest.d.ts`, but in that case jest global keywords (`it`, `describe` etc.) are lost for me.

I could make it work by defining the declaration file in the `altair-api` root. Is it okay?